File: 394. Decode String.cpp
class Solution {
public:
    // two stacks: O(N) time & space, 15.34%, ok
    string decodeString(string s) {
        stack<int> cnt;
        stack<string> str;
        string ret = "";
        for(int i = 0; i < s.size(); i++){
            if(isdigit(s[i])){
                int num = s[i++] - '0';
                while(i < s.size() && isdigit(s[i]))
                    num = num * 10 + (s[i++] - '0');
                i--;
                cnt.push(num);
            }
            else if(s[i] == '['){
                str.push(ret);  // key
                ret = "";
            }
            else if(s[i] == ']'){
                int num = cnt.top();
                string temp = str.top();
                cnt.pop();
                str.pop();
                while(num--) temp += ret;
                ret = temp;     // key
            }
            else ret += s[i];
        }
        return ret;
    }

    // recursive: O(N) time & space, 15.34%, ok
    string decodeString(string s) {
        int i = 0;
        return decodeString(s, i);
    }
    // bug: "int i", should return referrence, o/w hard to trace the index
    string decodeString(string s, int& i) {
        string ret = "";
        while(i < s.size() && s[i] != ']'){
            if(!isdigit(s[i])) ret += s[i++];
            else{
                int num = 0;
                while(i < s.size() && isdigit(s[i]))
                    num = num * 10 + (s[i++] - '0');
                i++; // '['
                string str = decodeString(s, i);
                i ++; // ']'
                while(num--) ret += str;
            }
        }
        return ret;
    }
};

File: 394. Decode String.py
class Solution(object):
    
    # brute force recursion: O(N) time & space
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        ret, i = '', 0
        while i < len(s):
            if not s[i].isdigit():
                while i < len(s) and not s[i].isdigit():
                    ret, i = ret + s[i], i + 1
            else:
                num = ''
                while s[i].isdigit(): 
                    num, i = num + s[i], i + 1   # bug: num can be multiple digits
                substr, num, i, left_brackets = s[i+1], int(num), i+2, 1
                while left_brackets != 0:
                    substr += s[i]
                    if s[i] == '[': left_brackets += 1
                    if s[i] == ']': left_brackets -= 1
                    i += 1
                ret += num * self.decodeString(substr[:-1])
        return ret 
    
    # stack !!!!: O(N) time & space
    def decodeString(self, s):
        stack, num = [["", 1]], ''
        for c in s:
            if c.isdigit():
                num += c
            elif c == '[': 
                stack.append(["", int(num)])
                num = ""
            elif c == ']': 
                substr, k = stack.pop()
                stack[-1][0] += substr * k
            else:
                stack[-1][0] += c
            print stack
        return stack[0][0]
        